core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,412 @@
+	}
+	return
+}
+
+// FetchGasPrice will use eth_gasPrice to fetch and cache the latest gas price from the RPC.
+func (u *UniversalEstimator) FetchGasPrice() (*assets.Wei, error) {
+	ctx, cancel := u.stopCh.CtxCancel(evmclient.ContextWithDefaultTimeout())
+	defer cancel()
+
+	gasPrice, err := u.client.SuggestGasPrice(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch gas price: %s", err)
+	}
+
+	promUniversalEstimatorGasPrice.WithLabelValues(u.chainID.String()).Set(float64(gasPrice.Int64()))
+
+	gasPriceWei := (*assets.Wei)(gasPrice)
+
+	u.logger.Debugf("fetched new gas price: %v", gasPriceWei)
+
+	u.gasPriceMu.Lock()
+	defer u.gasPriceMu.Unlock()
+	u.gasPrice = gasPriceWei
+	return u.gasPrice, nil
+}
+
+func (u *UniversalEstimator) getGasPrice() (*assets.Wei, error) {
+	u.gasPriceMu.RLock()
+	defer u.gasPriceMu.RUnlock()
+	if u.gasPrice == nil {
+		return u.gasPrice, fmt.Errorf("gas price not set")
+	}
+	return u.gasPrice, nil
+}
+
+// GetDynamicFee will fetch the cached dynamic prices.
+func (u *UniversalEstimator) GetDynamicFee(ctx context.Context, maxPrice *assets.Wei) (fee DynamicFee, err error) {
+	if fee, err = u.getDynamicPrice(); err != nil {
+		return
+	}
+
+	if fee.FeeCap == nil || fee.TipCap == nil {
+		return fee, fmt.Errorf("dynamic price not set")
+	}
+	if fee.FeeCap.Cmp(maxPrice) > 0 {
+		u.logger.Warnf("estimated maxFeePerGas: %v is greater than the maximum price configured: %v, returning the maximum price instead.",
+			fee.FeeCap, maxPrice)
+		fee.FeeCap = maxPrice
+		if fee.TipCap.Cmp(maxPrice) > 0 {
+			u.logger.Warnf("estimated maxPriorityFeePerGas: %v is greater than the maximum price configured: %v, returning the maximum price instead.",
+				fee.TipCap, maxPrice)
+			fee.TipCap = maxPrice
+		}
+	}
+
+	return
+}
+
+// FetchDynamicPrice uses eth_feeHistory to fetch the basFee of the latest block and the Nth maxPriorityFeePerGas percentiles
+// of the past X blocks. It also fetches the highest Zth maxPriorityFeePerGas percentile of the past X blocks. Z is configurable
+// and it represents the highest percentile we're willing to pay.
+// A buffer is added on top of the latest basFee to catch fluctuations in the next blocks. On Ethereum the increase is baseFee*1.125 per block.
+func (u *UniversalEstimator) FetchDynamicPrice() (fee DynamicFee, err error) {
+	ctx, cancel := u.stopCh.CtxCancel(evmclient.ContextWithDefaultTimeout())
+	defer cancel()
+
+	if u.config.BlockHistorySize == 0 {
+		return fee, fmt.Errorf("BlockHistorySize cannot be 0")
+	}
+	// RewardPercentile will be used for maxPriorityFeePerGas estimations and connectivityPercentile to set the highest threshold for bumping.
+	feeHistory, err := u.client.FeeHistory(ctx, u.config.BlockHistorySize, []float64{u.config.RewardPercentile, ConnectivityPercentile})
+	if err != nil {
+		return fee, fmt.Errorf("failed to fetch dynamic prices: %s", err)
+	}
+
+	// Latest base fee
+	baseFee := (*assets.Wei)(feeHistory.BaseFee[len(feeHistory.BaseFee)-1])
```

Comment:
nit: for the same reason as a previous comment
```suggestion
	baseFee := assets.NewWei(feeHistory.BaseFee[len(feeHistory.BaseFee)-1])

Reply:
Realized this is in a few different places below so I won't comment at every spot